### PR TITLE
feature: RFC3021 (/31) and /32 support added to snmp-scan.php

### PR DESCRIPTION
**Pull request by iviemeister**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Support RFC3021 - Add /31 and /32 subnet support to snmp-scan.php, add force-scan options for network and broadcast addresses, improve debugging messages.

### CODE REVIEW

File: snmp-scan.php [hunk 1]
```diff
@@ -46,62 +46,131 @@
 
 function perform_snmp_scan($net)
 {
-    global $stats, $config, $debug, $vdebug;
+    global $stats, $config, $debug, $vdebug, $force_network, $force_broadcast;
```
Comment (laf) [inline snmp-scan.php]: You should probably pass these in to the function rather than using global.

File: snmp-scan.php [hunk 2]
```diff
@@ -46,62 +46,131 @@
...
     echo 'Range: '.$net->network.'/'.$net->bitmask.PHP_EOL;
     $config['snmp']['timeout'] = 1;
     $config['snmp']['retries'] = 0;
     $config['fping_options']['retries']  = 0;
     $start = ip2long($net->network);
     $end   = ip2long($net->broadcast)-1;
+
+    if ($force_network == true) { //Force-scan network address
```
Comment (laf) [inline snmp-scan.php]: Needs to be ===

File: snmp-scan.php [hunk 3]
```diff
@@ -46,62 +46,131 @@
...
+        if ($debug == true) {
```
Comment (laf) [inline snmp-scan.php]: Please use d_echo() function instead of `if ($debug) { echo ''; }`

File: snmp-scan.php [hunk 4]
```diff
@@ -46,62 +46,131 @@
...
+            echo "Forcing network address scan".PHP_EOL;
+        }
+        $start = $start-1;
+    }
+
+    if ($force_broadcast == true) { //Force-scan broadcast address
```
Comment (laf) [inline snmp-scan.php]: All == true should be === to be explicit.

File: snmp-scan.php [hunk 5]
```diff
@@ -46,62 +46,131 @@
...
+        if ($debug == true) {
```
Comment (laf) [inline snmp-scan.php]: Same comment here for d_echo()

File: snmp-scan.php [hunk 6]
```diff
@@ -46,62 +46,131 @@
...
+            echo "Forcing broadcast address scan".PHP_EOL;
+        }
+        $end = $end+1;
+    }
+
+    if ($net->bitmask == 31) { //Handle RFC3021 /31 prefixes
```
Comment (laf) [inline snmp-scan.php]: Can use === here as well.

File: snmp-scan.php [hunk 7]
```diff
@@ -46,62 +46,131 @@
...
+        $start = ip2long($net->network)-1;
+        $end   = ip2long($net->broadcast);
+    }
+
+    if ($net->bitmask == 32) { //Handle single-host /32 prefixes
+        $start = ip2long($net->network)-1;
+        $end   = $start+1;
+    }
+
+    if ($debug == true) {
```
Comment (laf) [inline snmp-scan.php]: Can remove the if $debug check and just wrap the echo's in d_echo instead.

File: snmp-scan.php [hunk 8]
```diff
@@ -46,62 +46,131 @@
...
+            echo 'Network:   '.($net->network).PHP_EOL;
+            echo 'Broadcast: '.($net->broadcast).PHP_EOL.PHP_EOL;
+        }
+    }
+
     while ($start++ < $end) {
         $stats['count']++;
         $host = long2ip($start);
+
+        if ($vdebug == true) {
```
Comment (laf) [inline snmp-scan.php]: d_echo()

File: snmp-scan.php [hunk 9]
```diff
@@ -46,62 +46,131 @@
...
+            echo "Scanning: ".$host.PHP_EOL;
+        }
+
         if (match_network($config['autodiscovery']['nets-exclude'], $host)) {
-            echo '|';
+            if ($vdebug == true) {
```
Comment (laf) [inline snmp-scan.php]: d_echo()

Comment (laf) [inline snmp-scan.php]: ===

File: snmp-scan.php [hunk 10]
```diff
@@ -46,62 +46,131 @@
...
+
+    if ($net->bitmask == 32) { //Handle single-host /32 prefixes
+        $start = ip2long($net->network)-1;
+        $end   = $start+1;
+    }
+
+    if ($debug == true) {
+        if ($net->bitmask == 31) {
+            echo "RFC3021 network, hosts ".long2ip($start+1)." and ".long2ip($end).PHP_EOL.PHP_EOL;
+        } elseif ($net->bitmask == 32) {
```
Comment (laf) [inline snmp-scan.php]: `===`

File: snmp-scan.php [hunk 11]
```diff
@@ -46,62 +46,131 @@
...
+        $end   = ip2long($net->broadcast);
+    }
+
+    if ($net->bitmask == 32) { //Handle single-host /32 prefixes
+        $start = ip2long($net->network)-1;
+        $end   = $start+1;
+    }
+
+    if ($debug == true) {
+        if ($net->bitmask == 31) {
```
Comment (laf) [inline snmp-scan.php]: `===`

File: snmp-scan.php (hunk 8 above)
Comment (laf) [inline snmp-scan.php]: `===`

File: snmp-scan.php [hunk 12]
```diff
@@ -46,62 +46,131 @@
...
+                echo "Excluded by config.php".PHP_EOL.PHP_EOL;
+            } else {
+                echo '|';
+            }
             continue;
         }
         $test = isPingable($host);
         if ($test['result'] === false) {
-            echo '.';
+            if ($vdebug == true) {
```
Comment (laf) [inline snmp-scan.php]: Remove if ($vdebug) and wrap echo in d_echo()

File: snmp-scan.php [hunk 13]
```diff
@@ -46,62 +46,131 @@
...
+                echo "Unpingable Device".PHP_EOL.PHP_EOL;
+            } else {
+                echo '.';
+            }
             continue;
         }
         if (ip_exists($host)) {
             $stats['known']++;
-            echo '*';
+            if ($vdebug == true) {
```
Comment (laf) [inline snmp-scan.php]: Removed $vdebug and use d_echo()

**Comment by laf** [comment]
Thanks for the contribution. I've put some inline comments in place but stopped near the end, those comments apply all the way through though.

`===` should be used for strict checking.

`d_echo()` is available to use instead of checking if debug is set then echoing out.

**Comment by iviemeister** [comment]
Switched all == to ===, switched all "if ($debug..." to d_echo().

Left "if ($vdebug..." in scan loop, as those actually change output, not just add, and only in verbose debug mode.

**Comment by laf** [comment]
This has been merged now but $vdebug isn't for showing additional output (it's badly named really). It was put in to show the masked information we have in debug which hides sensitive info so really this still should be in d_echo()

**Comment by f0o** [comment]
Darn, i can do a follow-up patch

**Pull request merged**